### PR TITLE
Check once for Python modules

### PR DESCRIPTION
**Pull request by quantumsteve**
Please review the [developer documentation](https://github.com/QMCPACK/qmcpack/wiki/Development-workflow)
on the wiki of this project that contains help and requirements.

## Proposed changes

Describe what this PR changes and why.  If it closes an issue, link to it here
with [a supported keyword](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword).

The change caches the result of `TEST_PYTHON_MODULE` so that it is only called once for each Python module. This reduces the configuration time by almost 10 seconds.

`time cmake -G Ninja -DCMAKE_C_COMPILER=mpicc -DCMAKE_CXX_COMPILER=mpicxx ..`

`develop`
```
real	0m47.877s
user	1m26.576s
sys	1m45.337s
```
`check_python_once`
```
real	0m38.128s
user	0m36.330s
sys	0m31.351s
```

## What type(s) of changes does this code introduce?
_Delete the items that do not apply_

- Refactoring (no functional changes, no api changes)
- Build related changes

### Does this introduce a breaking change?

- No

## What systems has this change been tested on?

Ubuntu 20.04

## Checklist

_Update the following with a yes where the items apply. If you're unsure about any of them, don't hesitate to ask.  This is
simply a reminder of what we are going to look for before merging your code._

- Yes. This PR is up to date with current the current state of 'develop'
- n/a. Code added or changed in the PR has been clang-formatted
- n/a. This PR adds tests to cover any new code, or to catch a bug that is being fixed
- n/a. Documentation has been added (if appropriate)

### CODE REVIEW

**Comment by prckent** [comment]
Super. A net 4 line change for -10 seconds on the configure.

1. Did you also verify with make? (I assume the same, but that is an assumption)
2. Multiple make/ninja invocations (as for iterative development when one file is changed)?

**Comment by quantumsteve** [comment]
1. `time cmake -DCMAKE_C_COMPILER=mpicc -DCMAKE_CXX_COMPILER=mpicxx ..`
develop
```
real	0m49.518s
user	1m25.176s
sys	1m41.262s
```
check_python_once
```
real	0m40.012s
user	0m37.993s
sys	0m32.485s
```

**Comment by quantumsteve** [comment]
2. `time cmake .`
develop
```
real	0m44.213s
user	1m22.868s
sys	1m42.247s
```
check_python_once
```
real	0m31.423s
user	0m22.118s
sys	0m14.067s
```

File: CMake/python.cmake [hunk 1]
```diff
@@ -21,8 +21,12 @@ ENDFUNCTION()
 FUNCTION(CHECK_PYTHON_REQS module_list test_name add_test)
   set(${add_test} true PARENT_SCOPE)
   foreach(python_module IN LISTS ${module_list})
-    TEST_PYTHON_MODULE(${python_module} has_python_module)
-    if (NOT(has_python_module))
+    string(TOUPPER ${python_module} module_uppercase)
+    set(cached_result HAS_${module_uppercase}_PYTHON_MODULE)
```
Comment (ye-luo) [inline CMake/python.cmake:25]: could you rename cached_result to cached_variable_name

File: CMake/python.cmake [hunk 2]
```diff
@@ -21,8 +21,12 @@ ENDFUNCTION()
...
+    if (NOT DEFINED ${cached_result})
+      TEST_PYTHON_MODULE(${python_module} ${cached_result})
+    endif ()
+    if (NOT(${cached_result}))
```
Comment (ye-luo) [inline CMake/python.cmake:29]: `(NOT ${cached_result})` to reduce `()`

File: CMake/python.cmake (hunk 1 above)
Comment (quantumsteve) [reply CMake/python.cmake]: done

File: CMake/python.cmake (hunk 2 above)
Comment (quantumsteve) [reply CMake/python.cmake]: done

**Review by ye-luo — APPROVED** [review]

**Comment by prckent** [comment]
Test this please

**Pull request merged**